Template Pipeline: various fixes for compliance tests

dylhunn opened the pull request:
See individual commits for details.

dylhunn committed:
refactor(compiler): Add an escape hatch to use non-unique const pool names

The template pipeline was producing slightly different names than TemplateDefinitionBuilder for defer deps functions. I have added a workaround in the name of backwards compatibility, to avoid suffixing the const pool function names.

dylhunn committed:
refactor(compiler): Update a user visible error to match TDB

This error about global event targets is user-facing; update it to match TDB's error verbatim.

dylhunn committed:
refactor(compiler): Fix self-closing element source spans in template pipeline

When an element is self-closing, it will cause an `element` instruction to be emitted (instead of `elementStart`/`elementEnd`). In that case, we should use map whole source span for the instruction, not just the starting span.

dylhunn committed:
refactor(compiler): Defer when consumes a variable (binding) slot

The template pipeline was previously not reserving a variable slot for the result of the `deferWhen` instruction, which caused the `defer when` feature to crash at runtime.

dylhunn committed:
refactor(compiler): Host attribute bindings should always be extracted into hostAttrs

Host attribute literal bindings should not result in an `attribute` update instruction.

dylhunn committed:
refactor(compiler): Repeaters may consume a variable for their empty block

Repeaters consume a variable slot only when they have an empty block.

dylhunn committed:
refactor(compiler): Support class and non-class attributes with the same name

Some elements may have multiple bindings with the same name. We should accept and emit them all, as long as they have different kinds.

Co-authored-by: Miles Malerba <mmalerba@users.noreply.github.com>

mmalerba commented on packages/compiler/src/template/pipeline/src/phases/attribute_extraction.ts:
```diff
@@ -106,10 +106,10 @@ function extractAttributeOp(
   }
 
-  let extractable = op.expression.isConstant();
+  let extractable = op.isTextAttribute || op.expression.isConstant();
```
shouldn't `op.expression` always be constant anyways if we're dealing with a text attr?

dylhunn replied:
Not necessarily! Host binding attributes whose values are const TS variables are considered text attributes for these purposes, but the AST node does not satisfy `isConstant()`. Perhaps we should rename `isTextAttribute`.

dylhunn replied:
(e.g. see `host_with_ts_expression_node.ts` for an example)

mmalerba commented on packages/compiler/src/template/pipeline/src/phases/const_collection.ts:
```diff
@@ -102,12 +102,22 @@ class ElementAttributes {
     return this.byKind.get(ir.BindingKind.I18n) ?? FLYWEIGHT_ARRAY;
   }
 
+  isKnown(kind: ir.BindingKind, name: string, value: o.Expression|null) {
```
nice! you can try pulling in my test for this here: https://github.com/angular/angular/commit/f96f0e0d7761a370ceda74ea5a3a29e7646fac27, it may pass after this fix

dylhunn replied:
I pulled in the test, but I actually had to make some other changes as well to get it passing. I added a commit with your test and the relevant fixes.

dylhunn replied:
Note also -- I decided to accept a slight difference because it's more correct. See the commit for details.

dylhunn committed:
refactor(compiler): Not all attribute values beginning with a colon are namespaced

TemplateDefinitionBuilder is apparently more careful about when it attempts to split namespaces in attribute values. However, we are doing this on style attributes, which might start with a single `:`. Rather than refactor our logic to only try to split namespaces in some cases, we can just add an option to make namespace splitting fail gracefully. We only use this option for attributes, not elements.

Note also: the compiled code for this, while "correct" is absolutely insane. Maybe we should consider fixing this, as a matter of principle.

dylhunn commented:
@mmalerba I have actually added yet one more commit, to fix the majority of failures in my TGP -- that way we have useful results for tomorrow.

dylhunn commented:
caretaker: g3 presubmit is clean

jessicajaniuk commented:
This PR was merged into the repository by commit c9879457f628d593680e82d500925fb57aa114b9.
